### PR TITLE
[graph_trainer] Use separate EP process groups for overlap

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stacked PRs:
 * #3328
 * #3325
 * #3363
 * #3362
 * __->__#3369

--- --- ---

### [graph_trainer] Use separate EP process groups for overlap

When EP all-to-alls use the same process group object as FSDP or TP collectives, they share a NCCL communicator/stream and can serialize communication that the EP overlap schedule is trying to expose.

Add isolate_ep_process_group_pass as a dedicated EP pass instead of mixing this policy into FSDP bucketing. The pass scans traced collectives, identifies EP process-group users by _c10d_functional.all_to_all_single.default nodes, and compares the resolved process-group object identity against non-EP collective process groups. If an EP process group is object-identical to a non-EP collective group, the pass creates one high-priority EP-only process group over the same EP ranks and rewrites all traced EP all-to-all calls that used the shared source group.

The matching contract is intentionally process-group based because graph-level metadata does not carry DeviceMesh objects. Rank-set equality is only a validator: two distinct process groups may have the same ranks but already have separate NCCL communicators, so rank-set equality alone is not a sufficient reason to rewrite EP. The replacement communicator is always constructed from the EP process group itself, preserving the EP group rank order used by all-to-all split semantics.

FSDP all-gathers continue to use the existing extra FSDP group from overlap_fsdp_ag_rs_pass. Generic MoE dispatch/combine traceback metadata and EP_token_exchange scheduling metadata are not required for EP process-group isolation; the scheduling pass still owns token-exchange wait ordering.

Test Plan:
- pytest -q torchtitan/experiments/graph_trainer/tests/test_passes.py::TestOverlapFsdpAgRsPass

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 1]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+    return _get_or_create_extra_pg(
+        source_pg_name,
+        _EXTRA_EP_PG_REGISTRY,
+        group_desc="ep_extra",
+        high_priority=True,
+    )
+
+
+def _pg_rank_set(pg_name: str) -> frozenset[int]:
+    import torch.distributed as dist
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: nit: looks like we importing locally in 3+ functions, consider moving to the top of the file.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 2]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+    return False
+
+
+def _is_ep_all_to_all(node: fx.Node) -> bool:
+    custom = node.meta.get("custom", {})
+    return (
+        node.op == "call_function"
+        and "all_to_all_single" in str(node.target)
+        and isinstance(custom, dict)
+        and custom.get("EP") in ("dispatch", "combine")
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: Who adds this metadata?

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 3]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+    Reassign FSDP all-gathers and overlapping EP all-to-alls to extra NCCL
+    process groups.
 
     Discovers all distinct FSDP PGs by inspecting the graph (e.g. one for
     FSDP, another for expert-FSDP), creates an extra NCCL PG over the same
     ranks for each (giving it a separate CUDA stream), and rewrites every
     all-gather to the corresponding extra PG. This separates all-gathers
     from reduce-scatters onto different streams, enabling AG/RS overlap in
-    backward.
+    backward. If an EP all-to-all resolves to the same process group as an FSDP
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: Would we ever want to enable EP A2A in a separate stream but not a separate stream for AG/RS ? Should this be a separate pass with its own enable toggle?

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 4]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+        if is_fsdp_reduce_scatter(node):
+            fsdp_rs_pg_names.add(node.args[3])
 
     if not source_pg_names:
         return gm
 
     pg_mapping: dict[str, str] = {
         pg: _get_or_create_extra_fsdp_pg(pg) for pg in source_pg_names
     }
+    ep_source_pg_names = OrderedSet(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: nit: Can avoid extra O(N) scan of graph nodes by populating ep_source_pg_names inside previous for loop

**Review by aditvenk — APPROVED** [review]

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: `custom.get("EP") in ("dispatch", "combine")`  this is not reliable. 

In some come case, I see the entire EP is annotated as "compute"

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 5]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+                )
+            return True
+    return False
+
+
+def _is_ep_all_to_all(node: fx.Node) -> bool:
+    custom = node.meta.get("custom", {})
+    return (
+        node.op == "call_function"
+        and "all_to_all_single" in str(node.target)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: let's do direct target match. 

node.target == torch.ops.c10d.all_to_all_single.default

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 6]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+                    f"{pg_name} and {candidate_pg_name}"
+                )
+            return True
+    return False
+
+
+def _is_ep_all_to_all(node: fx.Node) -> bool:
+    custom = node.meta.get("custom", {})
+    return (
+        node.op == "call_function"
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: node.op == "call_function" is redundant.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 7]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
-    backward.
+    backward. If an EP all-to-all resolves to the same process group as an FSDP
+    reduce-scatter, it receives its own extra high-priority PG constructed from
+    the EP group. Rank-set equality alone is not enough: two different process
+    groups can have the same ranks but already use separate NCCL communicators.
 
     No-op when the graph has no FSDP all-gathers. Must be applied BEFORE
     bucketing passes so bucketed all-gathers inherit the new PG name.
     """
     source_pg_names: OrderedSet[str] = OrderedSet()
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py:128]: are these fsdp_ag_pg_names? rename if so

Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py:128]: these looks like a mixed of fsdp ag and rs pgs?

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 8]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
...
+    custom = node.meta.get("custom", {})
+    return (
+        node.op == "call_function"
+        and "all_to_all_single" in str(node.target)
+        and isinstance(custom, dict)
+        and custom.get("EP") in ("dispatch", "combine")
+    )
+
+
 def overlap_fsdp_ag_rs_pass(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py:110]: update pass name to reflect EP change. 

overlap_collectives_pass? 
reassign_pg_for_overlap_collectives_pass?

File: torchtitan/experiments/graph_trainer/fsdp_passes.py [hunk 9]
```diff
@@ -135,72 +136,171 @@ def fsdp_reshard_after_fwd_pass(
 # Each NCCL PG gets its own CUDA stream, so the extra PG is what enables
 # AG/RS overlap in backward.
 _EXTRA_FSDP_PG_REGISTRY: dict[str, str] = {}
+_EXTRA_EP_PG_REGISTRY: dict[str, str] = {}
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/fsdp_passes.py]: feels like they can be merged?

**Review by SherlockNoMad — APPROVED** [review]
thanks! lgtm overall, just minor comments.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Moved the EP process-group logic into its own graph-trainer pass, so the local imports are no longer spread through the FSDP pass. The remaining distributed imports are contained in the PG resolution/creation helpers that need them.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Reworked this to identify EP process-group users by the traced `_c10d_functional.all_to_all_single.default` calls themselves, instead of relying on generic `custom["EP"]` phase metadata. That keeps PG isolation tied to the collectives issued through the EP mesh rather than to scheduler-only traceback labels.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 3 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Split this out into `isolate_ep_process_group_pass` with its own graph-trainer pass hook. It rewrites EP all-to-all collectives when their process group object is shared with non-EP collectives.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 4 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: The standalone pass now performs one graph scan to classify EP and non-EP process groups, then rewrites only the matching EP all-to-alls. There is no extra FSDP-specific second scan for EP source PG discovery.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 5 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Updated the predicate to match the actual `_c10d_functional.all_to_all_single.default` target directly. The pass no longer requires the `EP_token_exchange` scheduler annotation for PG isolation.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 6 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Kept the `node.op` guard only as an FX safety check before reading call-function arguments. The semantic selection is the direct c10d all-to-all target; PG isolation no longer depends on the token-exchange annotation.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 7 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py:128]: Renamed and generalized the collection to distinguish EP all-to-all PGs from non-EP collective PGs. The pass now handles FSDP AG/RS, TP collectives, and other non-EP collectives uniformly.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 8 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py:110]: Renamed the pass to `isolate_ep_process_group_pass`, which reflects the actual behavior: EP all-to-alls are moved to an EP-only process group when they otherwise share a PG with non-EP collectives.

File: torchtitan/experiments/graph_trainer/fsdp_passes.py (hunk 9 above)
Comment (sanketpurandare) [reply torchtitan/experiments/graph_trainer/fsdp_passes.py]: Kept the EP PG isolation separate from FSDP bucketing because it is not FSDP-specific. It can apply when EP shares a PG with FSDP, TP, or another non-EP collective source.
